prow/config/cache.go

```diff
@@ -0,0 +1,388 @@
+
+	if promisePending {
+		// A promise exists, BUT the wrapped value inside it (p.result) might
+		// not be written to yet by the thread that is actually resolving the
+		// promise.
+		//
+		// For now we just unlock the overall lruCache itself so that it can
+		// service other GetOrAdd() calls to it.
+		lruCache.Unlock()
+
+		// If the type is not a promise type, there's no need to wait and we can
+		// just return immediately with an error.
+		promise, ok = maybePromise.(*Promise)
+		if !ok {
+			return nil, fmt.Errorf("invalid cache entry type '%T', expected '*Promise'", maybePromise)
+		}
+
+		// Block until the first thread originally created this promise has
+		// finished resolving it. Then it's safe to return the resolved values
+		// of the promise below.
+		//
+		// If the original thread resolved the promise already a long time ago
+		// (by closing the "resolve" channel), then this receive instruction
+		// will finish immediately and we will not block at all.
+		<-promise.resolve
+	} else {
+		// No promise exists for this key. In other words, we are the first
+		// thread to ask for this key's value and so We have no choice but to
+		// construct the value ourselves (this call is expensive!) and add it to
+		// the cache.
+		//
+		// If there are other concurrent threads that call GetOrAdd() with the
+		// same key and value constructors, we force them to use the same value
+		// as us (so that they don't have to also all valConstructor()). We do
+		// this with the following algorithm:
+		//
+		//  1. immediately create a promise to construct the value
+		//  2. actually construct the value (expensive operation)
+		//  3. resolve the promise to alert all threads looking at the same promise
+		//     get the value from step 2.
+		//
+		// This mitigation strategy is a kind of "duplicate suppression", also
+		// called "request coalescing". The problem of multiple requests for the
+		// same cache entry is also called "cache stampede".
+
+		// Step 1
+		//
+		// Let other threads know about our promise to construct the value. We
+		// don't care if the underlying LRU cache had to evict an existing
+		// entry.
+		promise = &Promise{resolve: make(chan struct{})}
+		_ = lruCache.Add(key, promise)
+		lruCache.Unlock()
+
+		// Step 2
+		//
+		// Construct the value (expensive operation).
+		promise.val, promise.err = valConstructor()
+
+		// Step 3
+		//
+		// Broadcast to all watchers of this promise that it is ready to be read
+		// from (no data race!).
+		close(promise.resolve)
+
+		// If the value construction (expensive operation) failed, then we
+		// delete the cached entry so that we may attempt to re-try again in the
+		// future (instead of waiting for the LRUCache to evict it on its own
+		// over time).
+		//
+		// TODO: If our cache implementation supports a TTL mechanism, then we
+		// could just set that instead and let the cached entry to expire on its
+		// own.
+		if promise.err != nil {
+			lruCache.Lock()
```

Comment:
There is technically a race here since we have to relock the cache. The entry we want to remove may already have been deleted and may even have been replaced with a new valid value.
That race is unlikely to occur and the outcome would at worst be unnecessarily invalidating a cache entry so this trick seems worth while and safe enough.